### PR TITLE
Add named loggers: OptionsWnd

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This is part of the broken up PR #1481.

This PR implements the setting of the logger thresholds from the `OptionsWnd`.

It follows the core implementation #1529. 

***Merge Message***
Add UI to OptionsWnd to set individual named logger's thresholds.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
rebase would be helpful for reviewing changes

**Comment by geoffthemedio** [comment]
Is this the expected appearance of the options screen?
log_options

**Comment by LGM-Doyle** [comment]
Rebased.

Yes it is supposed to look like that, since in this PR none of the `verbose-logging`, `verbose-combat-logging` or `TRACE_EXECUTION` code has been converted into loggers.  This just adds that frame work to do the conversion in PR #1481.

I have found a bug that I have not yet fixed.

Setting the logger state from the OptionsWnd in the lobby does not work correctly.
I think that the problem is that the server will not accept changes to its logger state from a client that is not the host, so it is not accepting the startup logger state from the client which is not yet the host.  I will fix this and the changes can be tacked on to this PR or submitted as a separate fix, as you see fit.

**Comment by LGM-Doyle** [comment]
This PR is rebased to account for the recent `Message` changes.

I also added the fix to set the logging system state from the lobby.  I did this by waiting until the host id is assigned and then sending the logging configuration message to the server.

**Comment by dbenage-cx** [comment]
Noting so it is not inadvertently merged: needs rebase as `GG::Connect` was removed.

**Comment by LGM-Doyle** [comment]
Thanks for pointing that out.  It is now rebased.

File: server/ServerApp.cpp [hunk 1]
```diff
@@ -440,6 +441,33 @@ void ServerApp::HandleShutdownMessage(const Message& msg, PlayerConnectionPtr pl
...
+
+    SetLoggerThresholds(options);
+
+    // Forward the message to all the AIs
+    const auto relay_options_message = LoggerConfigMessage(Networking::INVALID_PLAYER_ID, options);
+    for (ServerNetworking::established_iterator players_it = m_networking.established_begin();
+         players_it != m_networking.established_end(); ++players_it)
+    {
+        if ((*players_it)->GetClientType() == Networking::CLIENT_TYPE_AI_PLAYER) {
+            DebugLogger() << "Forwarding logging thresholds to AI " << (*players_it)->PlayerID();
```
Comment (geoffthemedio) [inline server/ServerApp.cpp:464]: could perhaps filter what log setting messages are forwarded to the AIs, as they don't need to know about changes to the human client or server log settings...

File: util/Logger.cpp [hunk 2]
```diff
@@ -49,37 +49,72 @@ namespace {
 
     // Compile time constant pointers to constant char arrays.
     constexpr const char* const log_level_names[] = {"trace", "debug", "info", "warn", "error"};
+
+    std::stringstream ValidLogLevelWarning(const std::string& text){
```
Comment (geoffthemedio) [inline util/Logger.cpp]: space before {

Comment (geoffthemedio) [reply util/Logger.cpp]: Valid... or Invalid?

File: util/Logger.cpp [hunk 3]
```diff
@@ -49,37 +49,72 @@ namespace {
...
+        std::stringstream ss;
+        ss << "\"" << text <<"\" is not a valid log level. "
+           << "Valid levels are ";
+
+        for(int ii = static_cast<int>(LogLevel::min); ii <= static_cast<int>(LogLevel::max); ++ii) {
```
Comment (geoffthemedio) [inline util/Logger.cpp]: space before (

File: util/Logger.cpp [hunk 4]
```diff
@@ -49,37 +49,72 @@ namespace {
...
+        return it->second;
+
+    WarnLogger(log) << ValidLogLevelWarning(text).str();
     return LogLevel::debug;
 }
 
+std::unordered_map<std::string, LogLevel> ValidNameToLogLevel() {
+    std::unordered_map<std::string, LogLevel> retval{};
+
+    for(int ii = static_cast<int>(LogLevel::min); ii <= static_cast<int>(LogLevel::max); ++ii) {
```
Comment (geoffthemedio) [inline util/Logger.cpp]: space b4 (

File: util/Logger.cpp [hunk 5]
```diff
@@ -106,10 +141,14 @@ std::string DumpIndent()
...
+    std::string& LocalUnnamedLoggerIdentifier() {
         // Create default logger name as a static function variable to avoid static initialization fiasco
-        static std::string default_exec_logger_name;
-        return default_exec_logger_name;
+        static std::string unnamed_logger_identifier;
+        return unnamed_logger_identifier;
+    }
+
+    const std::string& DisplayName(const std::string& channel_name) {
+        return (channel_name.empty() ? LocalUnnamedLoggerIdentifier() : channel_name);
```
Comment (geoffthemedio) [inline util/Logger.cpp]: prefer {} on same line, after definition, for single-line functions

File: network/Message.cpp [hunk 6]
```diff
@@ -1093,3 +1110,29 @@ FO_COMMON_API void ExtractDispatchCombatLogsMessageData(
...
+        for (size_t ii = 0; ii < size; ++ii) {
+            std::string option;
+            std::string name;
+            LogLevel level;
+            ia >> BOOST_SERIALIZATION_NVP(option);
+            ia >> BOOST_SERIALIZATION_NVP(name);
+            ia >> BOOST_SERIALIZATION_NVP(level);
+            options.insert(std::make_tuple(option, name, level));
+        }
+    } catch(const std::exception& err) {
```
Comment (geoffthemedio) [inline network/Message.cpp]: space b4 (

File: default/stringtables/en.txt [hunk 7]
```diff
@@ -1344,6 +1344,29 @@ Sets UI tooltip pop-up delay for alternate pop-ups, in ms.
...
+# Sub section title for logger threshold per application.
+OPTIONS_DB_UI_LOGGER_THRESHOLD_PER_PROCESS
+Per process
+
+# Sub section title for logger threshold of individual sources
+OPTIONS_DB_UI_LOGGER_THRESHOLD_PER_SOURCE
+Per source
+
+OPTIONS_DB_UI_LOGGER_THRESHOLD_TOOLTIP
+'''Each log output file (freeorion.log, freeoriond.log and AI_x.log) collects log records from sources(sections of code).
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: space before (sections

File: default/stringtables/en.txt [hunk 8]
```diff
@@ -1344,6 +1344,29 @@ Sets UI tooltip pop-up delay for alternate pop-ups, in ms.
...
+Per process
+
+# Sub section title for logger threshold of individual sources
+OPTIONS_DB_UI_LOGGER_THRESHOLD_PER_SOURCE
+Per source
+
+OPTIONS_DB_UI_LOGGER_THRESHOLD_TOOLTIP
+'''Each log output file (freeorion.log, freeoriond.log and AI_x.log) collects log records from sources(sections of code). 
+Setting the threshold for a source will enable logs records of that level or higher for that source. 
+Each process client, server and ai has a default source and may have additional detailed sources.'''
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: I'd prefer Oxford commas

Comment (geoffthemedio) [reply default/stringtables/en.txt]: adding (eg. combat resolution logs) might clarify this

File: default/stringtables/en.txt [hunk 9]
```diff
@@ -1344,6 +1344,29 @@ Sets UI tooltip pop-up delay for alternate pop-ups, in ms.
...
+
+OPTIONS_DB_UI_LOGGER_THRESHOLD_PER_PROCESS_TOOLTIP
+'''Setting the threshold for a process will produce only log records of that level or higher for the default source.'''
+
+OPTIONS_DB_UI_LOGGER_THRESHOLD_PER_SOURCE_TOOLTIP
+'''Setting the threshold for a source will produce logs of that level or higher for that source.'''
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: for all processes?

File: default/stringtables/en.txt [hunk 10]
```diff
@@ -1344,6 +1344,29 @@ Sets UI tooltip pop-up delay for alternate pop-ups, in ms.
...
+OPTIONS_DB_UI_LOGGER_THRESHOLD_PER_SOURCE
+Per source
+
+OPTIONS_DB_UI_LOGGER_THRESHOLD_TOOLTIP
+'''Each log output file (freeorion.log, freeoriond.log and AI_x.log) collects log records from sources(sections of code). 
+Setting the threshold for a source will enable logs records of that level or higher for that source. 
+Each process client, server and ai has a default source and may have additional detailed sources.'''
+
+OPTIONS_DB_UI_LOGGER_THRESHOLD_PER_PROCESS_TOOLTIP
+'''Setting the threshold for a process will produce only log records of that level or higher for the default source.'''
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: but will not affect other sources?

File: default/stringtables/en.txt (hunk 7 above)
Comment (LGM-Doyle) [reply default/stringtables/en.txt]: Fixed.

File: default/stringtables/en.txt (hunk 8 above)
Comment (LGM-Doyle) [reply default/stringtables/en.txt]: Both fixed.

File: default/stringtables/en.txt (hunk 10 above)
Comment (LGM-Doyle) [reply default/stringtables/en.txt]: Fixed.

File: default/stringtables/en.txt (hunk 9 above)
Comment (LGM-Doyle) [reply default/stringtables/en.txt]: Fixed.

File: network/Message.cpp (hunk 6 above)
Comment (LGM-Doyle) [reply network/Message.cpp]: Fixed.

File: server/ServerApp.cpp (hunk 1 above)
Comment (LGM-Doyle) [reply server/ServerApp.cpp:464]: I did not make this change.

The underlying logging system won't configure the its process logger with information from any other process.

File: util/Logger.cpp (hunk 2 above)
Comment (LGM-Doyle) [reply util/Logger.cpp]: Both changed.

File: util/Logger.cpp (hunk 3 above)
Comment (LGM-Doyle) [reply util/Logger.cpp]: Fixed.

File: util/Logger.cpp (hunk 4 above)
Comment (LGM-Doyle) [reply util/Logger.cpp]: Fixed.

File: util/Logger.cpp (hunk 5 above)
Comment (LGM-Doyle) [reply util/Logger.cpp]: Fixed.

File: server/ServerApp.cpp (hunk 1 above)
Comment (geoffthemedio) [reply server/ServerApp.cpp:464]: my concern was just the extra spam in the AI logs with irrelevant other-process logging level changes

**Pull request merged**